### PR TITLE
Context for constructor error is outer context

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes https://github.com/lampepfl/dotty/issues/7709

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/SymDenotations.scala [hunk 1]
```diff
@@ -824,16 +824,16 @@ object SymDenotations {
...
-          if (whyNot != null) whyNot append str
+      def isProtectedAccessOK: Boolean =
+        def fail(str: => String): false =
+          if whyNot != null then whyNot.append(str)
           false
-        }
         val cls = owner.enclosingSubClass
-        if (!cls.exists)
+        if !cls.exists then
+          val encl = if !ctx.owner.isConstructor then ctx else ctx.outersIterator.dropWhile(_.owner.isConstructor).next()
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/core/SymDenotations.scala:833]: The code does not handle secondary constructors. Will something like the following work?

```suggestion
          val encl = if ctx.owner.isPrimaryConstructor then ctx.enclosingClass.owner.enclosingClass else ctx
```

Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/core/SymDenotations.scala]: The expressions are the same, as the next outer context which is not owned by a ctor will be the correct enclosing class; also, both tests will discover an enclosing package symbol. I prefer your suggestion, which avoids an iterator -- although I almost talked myself out of it: the average view is that we seem to be skipping our class for the enclosing class; though from a context perspective, we're really just pushing out to the true enclosing class. (The context of expressions to super constructors is always confusing.) However, it seems `isConstructor` works here, also for secondaries; `outersIterator` would just plow through a couple more contexts in that case.

This change is a smallish tweak, but a good opportunity to take a look at scala 3 contexts...

**Comment by som-snytt** [comment]
Dotty CI claims the test output does not match the check file, but it looks the same to me. At this point, I wish the test rig would just display the diff (as partest learned to do, eventually).

File: compiler/src/dotty/tools/dotc/core/SymDenotations.scala (hunk 1 above)
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/core/SymDenotations.scala]: Sorry, I wasn't clear. I meant that in a secondary constructor, we might not want to go to the outer class --- as the current class itself could extend the related class.

**Comment by smarter** [comment]
Running `testCompilation --update-checkfiles tests/neg/i7709.scala` should fix the checkfile.

File: compiler/src/dotty/tools/dotc/core/SymDenotations.scala (hunk 1 above)
Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/core/SymDenotations.scala]: It's obvious when you put it that way.

**Comment by som-snytt** [comment]
Thanks @smarter I've stuck up a post-it note.

The test passes locally. When I munge the test, the "actual output" in sbt doesn't show the `|` on the LHS of the message output, although it is present as usual in the `.check.out`. I did verify my sha is pushed.

Soon I'll push an update and see what happens! I can't wait!

**Comment by som-snytt** [comment]
Just `--update-checkfiles` and squash. Previous feedback had been addressed.

**Review by liufengyun — APPROVED** [review]
LGTM 🎉

**Comment by som-snytt** [comment]
I came back for `--update-checkfiles` because no autocomplete and `partest --update-check`.

**Pull request merged**